### PR TITLE
PI-3552: Add concatenate support for Ancilary Variables and Cell Measures

### PR DESCRIPTION
**Pull request by stephenworsley**
Addresses the concatenate half of #3552.

### CODE REVIEW

**Comment by stephenworsley** [comment]
Since this involves changing the `_as_defn()` method for cell measures and ancillary variables, this probably depends on the work done in #3378.

File: lib/iris/cube.py [hunk 1]
```diff
@@ -505,6 +510,16 @@ def concatenate_cube(self, check_aux_coords=True):
             is not applied to auxilliary coordinates that span the dimension
             the concatenation is occuring along. Defaults to True.
 
+        * check_cell_measures
```
Comment (pp-mo) [inline lib/iris/cube.py]: Tiny point, but this has copied a mis-spelling : 2 'r's in 'occurring'
Also above : only 1 'l' in in 'auxiliary'

However, ancillary seems to right everywhere !
Now I just need to stop saying cell methods when I mean cell measures...

File: lib/iris/cube.py [hunk 2]
```diff
@@ -505,6 +510,16 @@ def concatenate_cube(self, check_aux_coords=True):
             is not applied to auxilliary coordinates that span the dimension
```
Comment (pp-mo) [inline lib/iris/cube.py]:
```suggestion
            is not applied to auxiliary coordinates that span the dimension
```

File: lib/iris/cube.py [hunk 3]
```diff
@@ -505,6 +510,16 @@ def concatenate_cube(self, check_aux_coords=True):
...
             the concatenation is occuring along. Defaults to True.
 
+        * check_cell_measures
+            Checks the cell measures of the cubes match. This check
+            is not applied to cell measures that span the dimension
+            the concatenation is occuring along. Defaults to True.
```
Comment (pp-mo) [inline lib/iris/cube.py]:
```suggestion
            the concatenation is occurring along. Defaults to True.
```

These spellings are actually _easier_ as edit suggestions ..

File: lib/iris/cube.py [hunk 4]
```diff
@@ -505,6 +510,16 @@ def concatenate_cube(self, check_aux_coords=True):
             is not applied to auxilliary coordinates that span the dimension
             the concatenation is occuring along. Defaults to True.
```
Comment (pp-mo) [inline lib/iris/cube.py]:
```suggestion
            the concatenation is occurring along. Defaults to True.
```

File: lib/iris/cube.py (hunk 3 above)
Comment (pp-mo) [inline lib/iris/cube.py]:
```suggestion
            the concatenation is occurring along. Defaults to True.
```

These spellings are actually _easier_ as edit suggestions.
But you can't do it to a multi-line comment.

File: lib/iris/cube.py [hunk 5]
```diff
@@ -505,6 +510,16 @@ def concatenate_cube(self, check_aux_coords=True):
...
 
+        * check_cell_measures
+            Checks the cell measures of the cubes match. This check
+            is not applied to cell measures that span the dimension
+            the concatenation is occuring along. Defaults to True.
+
+        * check_ancils
+            Checks the ancillary variables of the cubes match. This check
+            is not applied to ancillary variables that span the dimension
+            the concatenation is occuring along. Defaults to True.
```
Comment (pp-mo) [inline lib/iris/cube.py]:
```suggestion
            the concatenation is occurring along. Defaults to True.
```

File: lib/iris/tests/test_concatenate.py [hunk 6]
```diff
@@ -658,6 +706,26 @@ def test_concat_2x2d_aux_xy_bounds(self):
         self.assertEqual(len(result), 1)
         self.assertEqual(result[0].shape, (2, 4))
 
+    def test_concat_2y2d_cell_measure_x_y_xy(self):
+        cubes = []
+        x = (0, 2)
+        cubes.append(_make_cube(x, (0, 4), 1, cell_measure="x,y,xy"))
+        cubes.append(_make_cube(x, (4, 6), 2, cell_measure="x,y,xy"))
+        result = concatenate(cubes)
+        # self.assertCML(result, ("concatenate", "concat_2y2d_cell_measure_x_y_xy.cml"))
```
Comment (pp-mo) [inline lib/iris/tests/test_concatenate.py]: I think this assertCML check are worth enabling..
(and more below)

File: lib/iris/tests/test_concatenate.py [hunk 7]
```diff
@@ -658,6 +706,26 @@ def test_concat_2x2d_aux_xy_bounds(self):
...
+        self.assertEqual(len(result), 1)
+        self.assertEqual(result[0].shape, (6, 2))
+
+    def test_concat_2y2d_ancil_x_y_xy(self):
+        cubes = []
+        x = (0, 2)
+        cubes.append(_make_cube(x, (0, 4), 1, ancil="x,y,xy"))
+        cubes.append(_make_cube(x, (4, 6), 2, ancil="x,y,xy"))
+        result = concatenate(cubes)
+        # self.assertCML(result, ("concatenate", "concat_2y2d_ancil_x_y_xy.cml"))
```
Comment (pp-mo) [inline lib/iris/tests/test_concatenate.py]: another "assertCML check worth enabling"

**Comment by pp-mo** [comment]
> bypass cml checking

That is neat !!

**Comment by pp-mo** [comment]
Thanks for patience @stephenworsley  !

**Pull request merged**